### ISSUE TITLE
Performance issue when math enabled

### ISSUE THREAD

**Issue by jiangyy** [MacDownApp/macdown#415]
I am working with large document that contains numerous math blocks. When I am typing, the editor panel seems to be frequently blocked when MacDown started to render, and the keypress events are queued and processed after rendering is completed. Though keyboard events are not lost, this kind of lag significantly impacts the experience of editing. Would you have a chance to move the rendering of HTML to another thread, such that the editor is never blocked?

-- Mac OS X 10.10 on Macbook Air Late 2013 (i5-4250U, 4GB memory)

**Comment by uranusjr** [comment]
Math blocks are currently rendered with JavaScript (using MathJax), so he short answer is no, because it is impossible to run it in a background thread AFAIK.

I’ve been looking into switching to KaTeX, which is a lot faster, and can be run in a headless environment But as of now it still lacks many basic features, and I am reluctant to do it just now.

The only solution now is to turn off math rendering (or just turn off preview rendering completely), I’m afraid.

**Comment by RoyiAvital** [comment]
Hi,
Could you please allow KaTeX in parallel to MathJax.
As MathJax has many features KaTeX lacks and I think that way one could chose speed or features.

Thank You.

**Comment by eliquious** [comment]
KaTeX support would be great! I love MacDown and I really don't want to look for other apps but working with large documents with math enabled leads to many seconds of waiting and it interrupts the writing process.

**Comment by franklinyu** [comment]
I'm thinking about "letting user choose between MathJax and KaTeX"? MathJax for full feature, and KaTeX for lightweight rendering.

**Comment by eliquious** [comment]
That would be great!

**Comment by uranusjr** [comment]
Should be fixed with the latest prerelease. (Local KaTeX support would be a separate issue.)

**Comment by RoyiAvital** [comment]
@uranusjr ,
Please don't drop MathJaX.
Having both KaTeX and MathJaX (User controlled) would be great but MathJaX is a must as KaTeX lacks many features.

**Comment by franklinyu** [comment]
@RoyiAvital What? There is no plan to drop MathJax right now. Latest release simply improved MathJax.

**Comment by RoyiAvital** [comment]
Please don't do it in the future as well.
There are features needed which only supported by MathJaX.

**Comment by uranusjr** [comment]
@RoyiAvital I know KaTeX is lacking (the maintainers said explicitly they do not aim for feature completeness for performance reasons). If we ever switch to KaTeX I’d still leave MathJax in as a fallback (when KaTeX fails). You can relax.
